src/Neo/SmartContract/ApplicationEngine.cs

```diff
@@ -612,11 +612,30 @@ public void SetState<T>(T state)
 
         public bool IsHardforkEnabled(Hardfork hardfork)
         {
-            if (PersistingBlock is null)
-                return true;
-            if (!ProtocolSettings.Hardforks.TryGetValue(hardfork, out uint height))
+            // Get all the known hardforks
+            var allHardforks = Enum.GetValues(typeof(Hardfork)).Cast<Hardfork>().ToList();
+            allHardforks.Sort(); // Ensure they are sorted in the order of their definition
+
+            // Case 1: If no hardfork is specified in the configuration, enable all hardforks by default
+            if (PersistingBlock is null || ProtocolSettings.Hardforks.Count == 0)
                 return true;
-            return PersistingBlock.Index >= height;
+
+            // If the hardfork is not specified in the configuration, check if it's a new hardfork. If yes, disable it.
+            if (!ProtocolSettings.Hardforks.ContainsKey(hardfork))
+            {
+                // If the index of the hardfork in the list of known hardforks is greater than the index of the last hardfork in the configuration, disable it.
+                if (allHardforks.IndexOf(hardfork) > allHardforks.IndexOf(ProtocolSettings.Hardforks.Keys.Last()))
```

Comment:
Sorry for the nitpicking, but what would be the result of `ProtocolSettings.Hardforks.Keys.Last()` in case of an unordered JSON config:
```
HF_C: 123
HF_B: 100
```

Technically, this config is valid, but shouldn't we sort the `ProtocolSettings.Hardforks.Keys` before accessing the `Last()`?

Reply:
However, if we disallow situations like
```
HF_D: 123
HF_B: 100
```
Then it's not a problem.